File: server/app/concurrency.py
import asyncio
from typing import Awaitable, Callable, Sequence, TypeVar

T = TypeVar("T")


async def map_with_concurrency(items: Sequence[T], limit: int, fn: Callable[[T, int], Awaitable[None]]) -> None:
    """Runs `fn` over `items` with at most `limit` in flight at once.

    Used anywhere we'd otherwise fire one GitHub/Claude call per item sequentially
    (slow) or all at once (rate-limit/resource risk).
    """
    next_index = 0
    lock = asyncio.Lock()

    async def worker() -> None:
        nonlocal next_index
        while True:
            async with lock:
                if next_index >= len(items):
                    return
                i = next_index
                next_index += 1
            await fn(items[i], i)

    worker_count = min(limit, len(items))
    await asyncio.gather(*(worker() for _ in range(worker_count)))

File: server_express_backup/src/concurrency.js
// Bounded-parallelism map: runs `fn` over `items` with at most `limit` in flight at
// once. Used anywhere we'd otherwise fire one GitHub/Claude call per item sequentially
// (slow) or all at once (rate-limit/resource risk) — see prs.js's PR-list refresh and
// reviews.js's batch review.
export async function mapWithConcurrency(items, limit, fn) {
  let next = 0;
  async function worker() {
    while (next < items.length) {
      const i = next++;
      await fn(items[i], i);
    }
  }
  await Promise.all(Array.from({ length: Math.min(limit, items.length) }, worker));
}
